src/dsl.rs

```diff
@@ -53,12 +53,12 @@ impl<F, TraceArgs> CircuitContext<F, TraceArgs> {
     }
 
     /// Exposes the first step instance value of a forward signal as public.
-    pub fn expose(&mut self, forward: Queriable<F>) {
-        match forward {
-            Queriable::Forward(forward_signal, _) => {
-                self.sc.expose(forward_signal);
+    pub fn expose(&mut self, queriable: Queriable<F>, offset: ExposeOffset) {
+        match queriable {
+            Queriable::Forward(..) | Queriable::Shared(..) => {
+                self.sc.exposed.push((queriable, offset));
             }
-            _ => panic!("Expected forward signal"),
+            _ => panic!("Expected either ForwardSignal or SharedSignal"),
         }
```

Comment:
Here in `expose` we should check for Forward and Share as you've done, and then call `expose` in the AST rather than directly pushing to `exposed` of AST.

In general, we want functions that modifies AST to remain in AST, so that DSL can do simple calls of AST functions. Let me know if this makes sense!

Reply:
@qwang98 good catch